### PR TITLE
[1.0] Limit sync-fetch-span by max-reversible-blocks if needed

### LINKED ISSUE #528: P2P: Limit requested blocks to avoid max-reversible-blocks
Currently, `sync-fetch-span` provides an upper limit on how many blocks ahead of applied blocks (`chain_head`) are allowed to be requested. As it is today, the maximum allowed number of unapplied blocks that can be requested/received is `2 * sync-fetch-span - 1`. A large `sync-fetch-span` or a small `max-reversible-blocks` option can cause a node to shutdown if the accumulated reversible blocks exceeds the `max-reversible-blocks`.

Limit `net_plugin` requested blocks to the minimum of `max-reversible-blocks - forkdb.size()` and `sync-fetch-span` to avoid hitting the `max-reversible-blocks` during P2P syncing.

`max-reversible-blocks` PR is https://github.com/AntelopeIO/spring/pull/493

### PR DESCRIPTION
**Pull request by heifner**
Example output:
```
warn  2024-08-16T12:56:55.440 net-7     net_plugin.cpp:2021           active_sync_fetch_sp ] sync-fetch-span 5000 restricted to 3270 by max-reversible-blocks 3600, fork_db_size 329
info  2024-08-16T12:56:55.440 net-7     net_plugin.cpp:2179           operator()           ] ["jungle4.cryptolions.io:9870 - 34355da" - 2 95.216.44.76:9870] requesting range 1480342 to 1483611, fhead 1480341, lib 1480012
```

Resolves #528

### CODE REVIEW

File: plugins/net_plugin/net_plugin.cpp [hunk 1]
```diff
@@ -1994,18 +1995,30 @@ namespace eosio {
...
       ,sync_peer_limit( sync_peer_limit )
+      ,max_reversible_blocks(max_reversible_blocks)
       ,sync_state(in_sync)
       ,min_blocks_distance(min_blocks_distance)
    {
    }
 
+   uint32_t sync_manager::active_sync_fetch_span() const {
+      auto fork_db_size = my_impl->chain_plug->chain().fork_db_size();
+      auto reversible_remaining = max_reversible_blocks - fork_db_size;
```
Comment (greg7mdp) [inline plugins/net_plugin/net_plugin.cpp:2013]: We shutdown when `fork_db_size() >= max_reversible_blocks`, so really we should have:
```suggestion
      auto reversible_remaining = max_reversible_blocks - fork_db_size - 1;
```
I think.

**Review by greg7mdp — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp [hunk 2]
```diff
@@ -2529,20 +2543,24 @@ namespace eosio {
                } else {
                   // use chain head instead of fork head so we do not get too far ahead of applied blocks
                   uint32_t head = my_impl->get_chain_head_num();
```
Comment (greg7mdp) [inline plugins/net_plugin/net_plugin.cpp:2545]: Isn't using chain_head here a problem, as fork_db head can be higher?

File: plugins/net_plugin/net_plugin.cpp (hunk 1 above)
Comment (linh2931) [reply plugins/net_plugin/net_plugin.cpp:2013]: Do we need to check `max_reversible_blocks > fork_db_size` before doing subtract?

File: plugins/net_plugin/net_plugin.cpp (hunk 2 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:2545]: `fork_db` `head` will always be >= `chain_head`. Therefore, `head + fetch_span`  < `forkdb head + fetch_span`.

Comment (greg7mdp) [reply plugins/net_plugin/net_plugin.cpp:2545]: exactly. But we are concerned about `forkdb head + fetch_span` becoming greater than `forkdb.size()`.

Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:2545]: Example:
```c++
chain_head = 1000
fetch_span = 2000
fork_db.size() = 3000
fork_db.head = 4000

sync received block (blk_num) = 3001

// blk_num < chain_head + fetch_span 
if (3001 < 1000 + 2000) // false
   request more

// blk_num < fork_db.head + fetch_span
if (3001 < 4000 + 2000) // true
   request more
```

Therefore, we are more constrained to "request more" by using chain head.

**Review by linh2931 — APPROVED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 2 above)
Comment (greg7mdp) [reply plugins/net_plugin/net_plugin.cpp]: Yes, thanks, I totally misread that!

File: plugins/net_plugin/net_plugin.cpp [hunk 3]
```diff
@@ -2527,22 +2546,26 @@ namespace eosio {
...
+                  if (blk_num >= sync_last_requested_num) {
+                     // do not allow to get too far ahead (sync_fetch_span) of chain head
+                     auto fetch_span = active_sync_fetch_span();
+                     // use chain head instead of fork head so we do not get too far ahead of applied blocks
+                     uint32_t head = my_impl->get_chain_head_num();
+                     if (blk_num < head + fetch_span) {
+                        // block was not applied, possibly because we already have the block
+                        fc_dlog(logger, "Requesting ${fs} blocks ahead, head: ${h} fhead ${fh} blk_num: ${bn} sync_next_expected_num ${nen} "
+                                        "sync_last_requested_num: ${lrn}, sync_last_requested_block: ${lrb}",
+                                ("fs", fetch_span)("h", my_impl->get_chain_head_num())("fh", my_impl->get_fork_head_num())
```
Comment (greg7mdp) [inline plugins/net_plugin/net_plugin.cpp:2558]:
```suggestion
                                ("fs", fetch_span)("h", head)("fh", my_impl->get_fork_head_num())
```

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: Limit sync-fetch-span by max-reversible-blocks if needed.
Note:end

**Pull request merged**